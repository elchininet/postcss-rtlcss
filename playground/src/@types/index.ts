export type NumberOrString = string | number;
export type NumberOrStringArray = NumberOrString[];

export interface ObjectProps<T> {
    [key: string]: T;
}