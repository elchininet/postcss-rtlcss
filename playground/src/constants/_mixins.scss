$small: 576px;       // Phones in landscape mode
$medium: 768px;      // Tablet devices
$large: 992px;       // Desktop devices
$extralarge: 1200px; // Large desktop devices

$transformParams: 0.25s ease-in-out;

@mixin media($size) {
    @if $size == small {
        @media(min-width: $small) {
            @content;
        }
    } @else if $size == medium {
        @media(min-width: $medium) {
            @content;
        }
    } @else if $size == large {
        @media(min-width: $large) {
            @content;
        }
    } @else {
        @media(min-width: $extralarge) {
            @content;
        }
    }
    
}

@mixin buildTransition($properties) {
    $props: ();
    @each $p in $properties {
        $props: append($props, $p $transformParams, comma);
    }
    transition: $props;
}