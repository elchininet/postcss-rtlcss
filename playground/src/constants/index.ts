export enum COLORS {
    action = '#18AEC9',
    active = '#026C7F',
    gray = '#DDD',
    gray_dark = '#555',
    gray_darkest = '#1E1E1E',
    gray_light = '#EFEFEF',
    white = '#FFF'
}