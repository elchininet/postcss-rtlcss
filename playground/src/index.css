html,
body {
    height: 100%;
    margin: 0;
    padding: 0;
}

body {
    font-family: Arial, Helvetica, sans-serif;
    overscroll-behavior-y: contain;
}

#app {
    height: 100%;
}