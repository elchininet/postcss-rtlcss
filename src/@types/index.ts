import { Rule, AtRule } from 'postcss';

export interface ObjectWithProps<T> {
    [key: string]: T;
}

export enum Mode {
    combined = 'combined',
    override = 'override'
}

export enum Source {
    ltr = 'ltr',
    rtl = 'rtl'
}

export enum Autorename {
    disabled = 'disabled',
    flexible = 'flexible',
    strict = 'strict'
}

export type ModeValues = keyof typeof Mode;
export type SourceValues = keyof typeof Source;
export type AutorenameValues = keyof typeof Autorename;

export type strings = string | string[];

interface StringMapOptions {
    scope: '*' | 'selector' | 'url';
    ignoreCase: boolean;
}

export interface StringMap {
    name: string;
    priority: number;
    search: strings;
    replace: strings;
    options: StringMapOptions;
}

export interface PluginStringMap {
    name?: string;
    search: strings;
    replace: strings;
}

export interface PluginOptions {
    mode?: ModeValues;
    ltrPrefix?: strings;
    rtlPrefix?: strings;
    bothPrefix?: strings;
    source?: SourceValues;
    processUrls?: boolean;
    processKeyFrames?: boolean;
    useCalc?: boolean;
    stringMap?: PluginStringMap[];
    autoRename?: Autorename;
    greedy?: boolean;
}

export interface PluginOptionsNormalized extends Omit<Required<PluginOptions>, 'stringMap'> {
    stringMap: StringMap[];
}

export interface RulesObject {
    rule: Rule;
    ruleLTR: Rule | null;
    ruleRTL: Rule | null;
    ruleBoth: Rule | null;
}

export interface AtRulesObject {
    atRuleParams: string;
    atRule: AtRule;
    atRuleFlipped: AtRule;
}

export interface AtRulesStringMapObject {
    name: string;
    nameFlipped: string;
}

export interface AtRulesStringMap {
    [key: string]: AtRulesStringMapObject;
}

export interface ShortHandsData {
    [key: string]: {
        overridden: string | null;
        overrides: string[];
    };
}