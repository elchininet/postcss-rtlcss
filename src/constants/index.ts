export const COMMENT_TYPE = 'comment';
export const DECLARATION_TYPE = 'decl';
export const RULE_TYPE = 'rule';
export const AT_RULE_TYPE = 'atrule';
export const STRING_TYPE = 'string';
export const BOOLEAN_TYPE = 'boolean';
export const KEYFRAMES_NAME = 'keyframes';
export const ANIMATION_PROP = 'animation';
export const ANIMATION_NAME_PROP = 'animation-name';
export const RTL_COMMENT_REGEXP = /rtl:/;
export const RTL_CONTROL_DIRECTIVE_REG_EXP = /^\/\*!? *rtl:?(\w+)?:(\w+):?([^*]*?) *\*\/$/;
export const FLIP_PROPERTY_REGEXP = /(right|left)/i;

export enum CONTROL_DIRECTIVE {
    IGNORE = 'ignore'
}

export enum CONTROL_DIRECTIVE_BLOCK {
    BEGIN = 'begin',
    END = 'end'
}