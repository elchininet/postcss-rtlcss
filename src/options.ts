import { Mode, Source } from '@types';

export { Mode, Source };