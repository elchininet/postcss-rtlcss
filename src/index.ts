import postcss, { Root, Transformer } from 'postcss';
import { PluginOptions, RulesObject } from '@types';
import { parseOptions } from '@utilities/options';
import { parseRules } from '@rules/parser';
import { appendRules } from '@utilities/rules';
export { PluginOptions, Mode, Source, PluginStringMap,  } from '@types';

const transformer = (options: PluginOptions = {}): Transformer  => (
    (css: Root): void => {
        const optionsParsed = parseOptions(options);
        const result: RulesObject[] = parseRules(css, optionsParsed);
        appendRules(result);           
    }
);

export const postcssRTLCSS = postcss.plugin<PluginOptions>('postcss-rtlcss', transformer);