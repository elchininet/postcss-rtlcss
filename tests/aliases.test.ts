import postcss from 'postcss';
import postcssRTLCSS from '../src';
import { PluginOptions } from '../src/@types';
import { readCSSFile, runTests } from './utils';
import { aliases } from './constants';

const baseOptions: PluginOptions = { aliases };

runTests(baseOptions, (pluginOptions: PluginOptions): void => {

  describe(`[[Mode: ${pluginOptions.mode}]] Aliases Tests:`, (): void => {

    let input = '';
  
    beforeEach(async (): Promise<void> => {
      input = input || await readCSSFile('input-variables.css');
    });
  
    it('aliases default', (): void => {
      const options: PluginOptions = { ...pluginOptions, aliases: {} };
      const output = postcss([postcssRTLCSS(options)]).process(input);
      expect(output.css).toMatchSnapshot();
      expect(output.warnings()).toHaveLength(0);
    });
  
    it('aliases map', (): void => {
      const options: PluginOptions = { ...pluginOptions };
      const output = postcss([postcssRTLCSS(options)]).process(input);
      expect(output.css).toMatchSnapshot();
      expect(output.warnings()).toHaveLength(0);
    });
  
    it('wrong aliases', (): void => {
      const options: PluginOptions = { ...pluginOptions, aliases: {base: true, parse: false} as unknown as Record<string, string> };
      const output = postcss([postcssRTLCSS(options)]).process(input);
      expect(output.css).toMatchSnapshot();
      expect(output.warnings()).toHaveLength(0);
    });
  
  });

});