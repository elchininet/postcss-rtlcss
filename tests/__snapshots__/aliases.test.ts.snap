// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[Mode: combined]] Aliases Tests: aliases default 1`] = `
":root {
    --small-padding: 2px 4px 8px 16px;
    --large-padding: 4px 8px 16px 32px;
    --custom-margin: 2px;
    --small-margin: 2px 4px 8px 16px;
    --large-margin: 4px 8px 16px 32px;
}

.test1 {
    margin: var(--small-margin);
    padding: var(--small-padding);
}

.test1.large {
    margin: var(--large-margin);
    padding: var(--large-padding);
}

[dir=\\"ltr\\"] .test1.large {
    left: 10px;
}

[dir=\\"rtl\\"] .test1.large {
    right: 10px;
}

.test2 {
    margin: var(--custom-margin);
}"
`;

exports[`[[Mode: combined]] Aliases Tests: aliases map 1`] = `
":root {
    --custom-margin: 2px;
}

[dir=\\"ltr\\"]:root {
    --small-padding: 2px 4px 8px 16px;
    --large-padding: 4px 8px 16px 32px;
    --small-margin: 2px 4px 8px 16px;
    --large-margin: 4px 8px 16px 32px;
}

[dir=\\"rtl\\"]:root {
    --small-padding: 2px 16px 8px 4px;
    --large-padding: 4px 32px 16px 8px;
    --small-margin: 2px 16px 8px 4px;
    --large-margin: 4px 32px 16px 8px;
}

.test1 {
    margin: var(--small-margin);
    padding: var(--small-padding);
}

.test1.large {
    margin: var(--large-margin);
    padding: var(--large-padding);
}

[dir=\\"ltr\\"] .test1.large {
    left: 10px;
}

[dir=\\"rtl\\"] .test1.large {
    right: 10px;
}

.test2 {
    margin: var(--custom-margin);
}"
`;

exports[`[[Mode: combined]] Aliases Tests: wrong aliases 1`] = `
":root {
    --small-padding: 2px 4px 8px 16px;
    --large-padding: 4px 8px 16px 32px;
    --custom-margin: 2px;
    --small-margin: 2px 4px 8px 16px;
    --large-margin: 4px 8px 16px 32px;
}

.test1 {
    margin: var(--small-margin);
    padding: var(--small-padding);
}

.test1.large {
    margin: var(--large-margin);
    padding: var(--large-padding);
}

[dir=\\"ltr\\"] .test1.large {
    left: 10px;
}

[dir=\\"rtl\\"] .test1.large {
    right: 10px;
}

.test2 {
    margin: var(--custom-margin);
}"
`;

exports[`[[Mode: override]] Aliases Tests: aliases default 1`] = `
":root {
    --small-padding: 2px 4px 8px 16px;
    --large-padding: 4px 8px 16px 32px;
    --custom-margin: 2px;
    --small-margin: 2px 4px 8px 16px;
    --large-margin: 4px 8px 16px 32px;
}

.test1 {
    margin: var(--small-margin);
    padding: var(--small-padding);
}

.test1.large {
    margin: var(--large-margin);
    padding: var(--large-padding);
    left: 10px;
}

[dir=\\"rtl\\"] .test1.large {
    left: auto;
    right: 10px;
}

.test2 {
    margin: var(--custom-margin);
}"
`;

exports[`[[Mode: override]] Aliases Tests: aliases map 1`] = `
":root {
    --small-padding: 2px 4px 8px 16px;
    --large-padding: 4px 8px 16px 32px;
    --custom-margin: 2px;
    --small-margin: 2px 4px 8px 16px;
    --large-margin: 4px 8px 16px 32px;
}

[dir=\\"rtl\\"]:root {
    --small-padding: 2px 16px 8px 4px;
    --large-padding: 4px 32px 16px 8px;
    --small-margin: 2px 16px 8px 4px;
    --large-margin: 4px 32px 16px 8px;
}

.test1 {
    margin: var(--small-margin);
    padding: var(--small-padding);
}

.test1.large {
    margin: var(--large-margin);
    padding: var(--large-padding);
    left: 10px;
}

[dir=\\"rtl\\"] .test1.large {
    left: auto;
    right: 10px;
}

.test2 {
    margin: var(--custom-margin);
}"
`;

exports[`[[Mode: override]] Aliases Tests: wrong aliases 1`] = `
":root {
    --small-padding: 2px 4px 8px 16px;
    --large-padding: 4px 8px 16px 32px;
    --custom-margin: 2px;
    --small-margin: 2px 4px 8px 16px;
    --large-margin: 4px 8px 16px 32px;
}

.test1 {
    margin: var(--small-margin);
    padding: var(--small-padding);
}

.test1.large {
    margin: var(--large-margin);
    padding: var(--large-padding);
    left: 10px;
}

[dir=\\"rtl\\"] .test1.large {
    left: auto;
    right: 10px;
}

.test2 {
    margin: var(--custom-margin);
}"
`;
